### DISCUSSION TITLE
Is it possible to use PIL images in Toga?

### DISCUSSION THREAD

**Discussion by ego-lay-atman-bay** [beeware/toga#1986] [Q&A]
I am using a python package that heavily uses PIL for creating images. I would like to be able to use PIL images in Toga without having to save the image to a BytesIO object, and grab the bytes from that. It would probably speed up the app.

**Comment by freakboy3742** [comment]
Yes, it's possible. The [documentation of the Image class constructor](https://toga.readthedocs.io/en/stable/reference/api/resources/images.html#toga.images.Image) details the API, and the[image example](https://github.com/beeware/toga/tree/main/examples/imageview) in the Toga repository shows that API in use.

**Reply by mhsmith** [reply]
@freakboy3742 I'm not sure if you understood the question correctly. As far as I know, there is no better way to use a PIL image in Toga than saving it to a BytesIO object. 

@ego-lay-atman-bay However, PNG compression is very fast, so I doubt this is significantly slowing down your app. If you think otherwise, please provide some timing data.

**Reply by freakboy3742** [reply]
Ah, you're right - I didn't consider the native PIL Image class. 

I can see how this might be mildly more convenient as an end user, but it definitely wouldn't be faster, as any implementation would need to marshal from PIL to bytes anyway, as none of the platform backends speak PIL formats natively. 

It would also introduce a dependency on PIL into Toga, which is something that I'd rather avoid. It might be possible to manage this by *not* declaring a PIL as a dependency of Toga, and doing a conditional import... that way user code that uses PIL can bring in the dependency and Toga will accommodate that.

**Comment by ego-lay-atman-bay** [comment]
> It might be possible to manage this by _not_ declaring a PIL as a dependency of Toga, and doing a conditional import... that way user code that uses PIL can bring in the dependency and Toga will accommodate that.

That would probably work the best. For now, I guess I will save the image to a BytesIO object, and get the data.
